src/csharp/Grpc.Core/Internal/DefaultDeserializationContext.cs

```diff
@@ -46,10 +46,16 @@ public DefaultDeserializationContext()
 
         public override byte[] PayloadAsNewBuffer()
         {
-            var buffer = new byte[payloadLength];
-            FillContinguousBuffer(bufferReader, buffer);
+            var buffer = payloadLength == 0 ? EmptyByteArray : new byte[payloadLength];
+            if (payloadLength != 0 | bufferReader == null)
+            {   // ^^ this might look odd, but when empty, we *want* this to fail
+                // if this is because it has been reset and not assigned; we make
+                // it fail in a known way by trying to fill from the (invalid) source
```

Comment:
If we decide to update this PR, it would be good to add a test to DefaultSerializationContextTest to make sure the behavior is as expected.